### ISSUE TITLE
Bank page when i use a template  in printer GLPI 9.5.2 but work with GLPI 9.5 

### ISSUE THREAD

**Issue by thibautjacques** [glpi-project/glpi#8218]
**Bank page when i use a template in printer GLPI 9.5.2 but work with GLPI 9.5 **
When i use a template in printers (for add a printer in DB) 

i have no problem with other template  (exemple monitor or computer)
i m on IIS 

**Page(s) URL**

/glpi/front/printer.form.php

**To reproduce**

first create a Templates for a printer 
1. Go to '.Assets..'
2. Click on '.printer'
3. click on + for add printer 
5. click on your template 
6. add a data 
7. click on add
8. see the error  (website is bank)

for bypass
reload the website  ../glpi/front/printer.form.php
ok the printer is add in DB

  

**Expected behavior**

I have that error with all web browser  (chrome;firefox..;)
I have no that error with GLPI 9.5 version on IIS 
It's a clean version 9.5.2  no data no plugin
I just create a template and test 

When a use a  blank template , i have no error 

**Logs**
Php-error 

```
[2020-11-02 18:43:13] glpiphplog.WARNING: Test logger  
[2020-11-02 19:42:25] glpiphplog.CRITICAL:   *** Uncaught Exception RuntimeException: Cannot guess  in C:\inetpub\wwwroot\glpi\glpi\inc\commondbchild.class.php at line 867
  Backtrace :
  inc\features\clonable.class.php:68                 CommonDBChild::getItemField()
  inc\commondbtm.class.php:1234                      Printer->post_clone()
  inc\commondbtm.class.php:1070                      CommonDBTM->clone()
  front\printer.form.php:50                          CommonDBTM->add()

```

**Screenshots**

![image](https://user-images.githubusercontent.com/73845313/97899602-cd525b00-1d39-11eb-895a-cd94f49fabec.png)

![image](https://user-images.githubusercontent.com/73845313/97897612-063d0080-1d37-11eb-96f9-06bfcb2acf63.png)

![image](https://user-images.githubusercontent.com/73845313/97897646-11902c00-1d37-11eb-9347-5bf40efd977d.png)

![image](https://user-images.githubusercontent.com/73845313/97897665-181ea380-1d37-11eb-8c33-16975090c378.png)

**Your GLPI setup (you can find it in Setup > General menu, System tab)**

```
[code]
 
GLPI 9.5.2 (/glpi => C:\inetpub\wwwroot\glpi\glpi)
Installation mode: TARBALL

Server
 
Operating system: Windows NT GLPI 10.0 build 17763 (Windows Server 2016) AMD64
PHP 7.4.1 cgi-fcgi (Core, PDO, Phar, Reflection, SPL, SimpleXML, bcmath, bz2, calendar, cgi-fcgi, ctype, curl, date, dom, exif,
	fileinfo, filter, gd, gettext, hash, iconv, imap, intl, json, ldap, libxml, mbstring, mysqli, mysqlnd, openssl, pcre, pdo_mysql,
	pdo_sqlite, readline, session, snmp, soap, sodium, standard, tidy, tokenizer, xml, xmlreader, xmlrpc, xmlwriter, zip, zlib)
Setup: max_execution_time="300" memory_limit="128M" post_max_size="8M" safe_mode="" session.save_handler="files"
	upload_max_filesize="2M" 
Software: Microsoft-IIS/10.0
	Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4350.0 Iron Safari/537.36
Server Software: mariadb.org binary distribution
	Server Version: 10.5.4-MariaDB
	Server SQL Mode: 
	Parameters: root@localhost/glpi
	Host info: localhost via TCP/IP
	
PHP version is at least 7.2.0 - Perfect!PHP version is at least 7.2.0 - Perfect!
Sessions support is available - Perfect!Sessions support is available - Perfect!
Allocated memory > 64 Mio - Perfect!Allocated memory > 64 Mio - Perfect!
mysqli extension is installedmysqli extension is installed
ctype extension is installedctype extension is installed
fileinfo extension is installedfileinfo extension is installed
json extension is installedjson extension is installed
mbstring extension is installedmbstring extension is installed
iconv extension is installediconv extension is installed
zlib extension is installedzlib extension is installed
curl extension is installedcurl extension is installed
gd extension is installedgd extension is installed
simplexml extension is installedsimplexml extension is installed
intl extension is installedintl extension is installed
ldap extension is installedldap extension is installed
apcu extension is not presentapcu extension is not present
Zend OPcache extension is not presentZend OPcache extension is not present
xmlrpc extension is installedxmlrpc extension is installed
CAS extension is not presentCAS extension is not present
exif extension is installedexif extension is installed
zip extension is installedzip extension is installed
bz2 extension is installedbz2 extension is installed
sodium extension is installedsodium extension is installed
Database version seems correct (10.5.4) - Perfect!Database version seems correct (10.5.4) - Perfect!
Timezones seems not loaded, see https://glpi-install.readthedocs.io/en/latest/timezones.html.Timezones seems not loaded, see https://glpi-install.readthedocs.io/en/latest/timezones.html.
The log file has been created successfully.The log file has been created successfully.
Write access to C:\inetpub\wwwroot\glpi\glpi/config has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/config has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_dumps has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_dumps has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_sessions has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_sessions has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_cron has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_cron has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_graphs has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_graphs has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_lock has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_lock has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_plugins has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_plugins has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_tmp has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_tmp has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_cache has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_cache has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_rss has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_rss has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_uploads has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_uploads has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/files/_pictures has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/files/_pictures has been validated.
Write access to C:\inetpub\wwwroot\glpi\glpi/marketplace has been validated.Write access to C:\inetpub\wwwroot\glpi\glpi/marketplace has been validated.
Web access to the files directory should not be allowed but this cannot be checked automatically on this instance. Make sure access to error log file (/glpi/files/_log/php-errors.log) is forbidden; otherwise review .htaccess file and web server configuration.Web access to the files directory should not be allowed but this cannot be checked automatically on this instance.
Make sure access to error log file (/glpi/files/_log/php-errors.log) is forbidden; otherwise review .htaccess file and web server configuration.

GLPI constants
 
GLPI_ROOT: C:\inetpub\wwwroot\glpi\glpi
GLPI_CONFIG_DIR: C:\inetpub\wwwroot\glpi\glpi/config
GLPI_VAR_DIR: C:\inetpub\wwwroot\glpi\glpi/files
GLPI_MARKETPLACE_DIR: C:\inetpub\wwwroot\glpi\glpi/marketplace
GLPI_USE_CSRF_CHECK: 1
GLPI_CSRF_EXPIRES: 7200
GLPI_CSRF_MAX_TOKENS: 100
GLPI_TELEMETRY_URI: https://telemetry.glpi-project.org
GLPI_INSTALL_MODE: TARBALL
GLPI_NETWORK_MAIL: glpi@teclib.com
GLPI_NETWORK_SERVICES: https://services.glpi-network.com
GLPI_MARKETPLACE_PRERELEASES: 
GLPI_USER_AGENT_EXTRA_COMMENTS: 
GLPI_AJAX_DASHBOARD: 1
GLPI_CALDAV_IMPORT_STATE: 0
GLPI_DEMO_MODE: 0
GLPI_FORCE_EMPTY_SQL_MODE: 1
GLPI_DOC_DIR: C:\inetpub\wwwroot\glpi\glpi/files
GLPI_CACHE_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_cache
GLPI_CRON_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_cron
GLPI_DUMP_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_dumps
GLPI_GRAPH_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_graphs
GLPI_LOCAL_I18N_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_locales
GLPI_LOCK_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_lock
GLPI_LOG_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_log
GLPI_PICTURE_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_pictures
GLPI_PLUGIN_DOC_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_plugins
GLPI_RSS_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_rss
GLPI_SESSION_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_sessions
GLPI_TMP_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_tmp
GLPI_UPLOAD_DIR: C:\inetpub\wwwroot\glpi\glpi/files/_uploads
GLPI_NETWORK_REGISTRATION_API_URL: https://services.glpi-network.com/api/registration/
GLPI_MARKETPLACE_PLUGINS_API_URI: https://services.glpi-network.com/api/glpi-plugins/
GLPI_I18N_DIR: C:\inetpub\wwwroot\glpi\glpi/locales
GLPI_VERSION: 9.5.2
GLPI_SCHEMA_VERSION: 9.5.2
GLPI_MIN_PHP: 7.2.0
GLPI_YEAR: 2020

Libraries
 
htmlawed/htmlawed version 1.2.5 in (C:\inetpub\wwwroot\glpi\glpi\vendor\htmlawed\htmlawed)
phpmailer/phpmailer version 6.1.6 in (C:\inetpub\wwwroot\glpi\glpi\vendor\phpmailer\phpmailer\src)
simplepie/simplepie version 1.5.5 in (C:\inetpub\wwwroot\glpi\glpi\vendor\simplepie\simplepie\library)
tecnickcom/tcpdf version 6.3.5 in (C:\inetpub\wwwroot\glpi\glpi\vendor\tecnickcom\tcpdf)
michelf/php-markdown in (C:\inetpub\wwwroot\glpi\glpi\vendor\michelf\php-markdown\Michelf)
true/punycode in (C:\inetpub\wwwroot\glpi\glpi\vendor\true\punycode\src)
iamcal/lib_autolink in (C:\inetpub\wwwroot\glpi\glpi\vendor\iamcal\lib_autolink)
sabre/dav in (C:\inetpub\wwwroot\glpi\glpi\vendor\sabre\dav\lib\DAV)
sabre/http in (C:\inetpub\wwwroot\glpi\glpi\vendor\sabre\http\lib)
sabre/uri in (C:\inetpub\wwwroot\glpi\glpi\vendor\sabre\uri\lib)
sabre/vobject in (C:\inetpub\wwwroot\glpi\glpi\vendor\sabre\vobject\lib)
laminas/laminas-cache in (C:\inetpub\wwwroot\glpi\glpi\vendor\laminas\laminas-cache\src)
laminas/laminas-i18n in (C:\inetpub\wwwroot\glpi\glpi\vendor\laminas\laminas-i18n\src)
laminas/laminas-serializer in (C:\inetpub\wwwroot\glpi\glpi\vendor\laminas\laminas-serializer\src)
monolog/monolog in (C:\inetpub\wwwroot\glpi\glpi\vendor\monolog\monolog\src\Monolog)
sebastian/diff in (C:\inetpub\wwwroot\glpi\glpi\vendor\sebastian\diff\src)
elvanto/litemoji in (C:\inetpub\wwwroot\glpi\glpi\vendor\elvanto\litemoji\src)
symfony/console in (C:\inetpub\wwwroot\glpi\glpi\vendor\symfony\console)
scssphp/scssphp in (C:\inetpub\wwwroot\glpi\glpi\vendor\scssphp\scssphp\src)
laminas/laminas-mail in (C:\inetpub\wwwroot\glpi\glpi\vendor\laminas\laminas-mail\src\Protocol)
laminas/laminas-mime in (C:\inetpub\wwwroot\glpi\glpi\vendor\laminas\laminas-mime\src)
rlanvin/php-rrule in (C:\inetpub\wwwroot\glpi\glpi\vendor\rlanvin\php-rrule\src)
blueimp/jquery-file-upload in (C:\inetpub\wwwroot\glpi\glpi\vendor\blueimp\jquery-file-upload\server\php)
ramsey/uuid in (C:\inetpub\wwwroot\glpi\glpi\vendor\ramsey\uuid\src)
psr/log in (C:\inetpub\wwwroot\glpi\glpi\vendor\psr\log\Psr\Log)
psr/simple-cache in (C:\inetpub\wwwroot\glpi\glpi\vendor\psr\simple-cache\src)
mexitek/phpcolors in (C:\inetpub\wwwroot\glpi\glpi\vendor\mexitek\phpcolors\src\Mexitek\PHPColors)
guzzlehttp/guzzle in (C:\inetpub\wwwroot\glpi\glpi\vendor\guzzlehttp\guzzle\src)
guzzlehttp/psr7 in (C:\inetpub\wwwroot\glpi\glpi\vendor\guzzlehttp\psr7\src)
wapmorgan/unified-archive in (C:\inetpub\wwwroot\glpi\glpi\vendor\wapmorgan\unified-archive\src)
paragonie/sodium_compat in (C:\inetpub\wwwroot\glpi\glpi\vendor\paragonie\sodium_compat\src)

SQL replicas
 
Not active

Notifications
 
Way of sending emails: PHP

Mails receivers
 

Plugins list
 

[/code]
A copier/coller dans vos demandes d'assistance
```

**Additional context**

I have no error if i do the same action  with GLPI 9.5
(add a printer with a template)  
Warning no a blank template , just a template with  data 

thank

**Comment by cconard96** [comment]
Probably the same issue as #8108 since templates and cloning share some of the same code now. I didn't get a chance to look into it much, but one or more of the relations was failing to clone. If you don't have items linked, you can't tell they failed because everything else looks OK beyond getting the blank page.

**Comment by cedric-anne** [comment]
Well, I confirm the bug exists, whatever the template contains. Indeed, exception is thrown on related items computation, so even if template has no related items, exception will happen.

Reason is that `Cartridge` class is a relation between `Printer` and `CartridgeItem`, but extends `CommonDBChild`.
Applying following patch prevent this issue, but I am not sure that has no side-effects.
```diff
diff --git a/inc/cartridge.class.php b/inc/cartridge.class.php
index 6cfb88a8c3..311adf0ccf 100644
--- a/inc/cartridge.class.php
+++ b/inc/cartridge.class.php
@@ -40,7 +40,7 @@ if (!defined('GLPI_ROOT')) {
  * @see CartridgeItem
  * @author Julien Dombre
  **/
-class Cartridge extends CommonDBChild {
+class Cartridge extends CommonDBRelation {
    use Glpi\Features\Clonable;
 
    // From CommonDBTM
@@ -48,9 +48,11 @@ class Cartridge extends CommonDBChild {
    public $dohistory                   = true;
    public $no_form_page                = true;
 
-   // From CommonDBChild
-   static public $itemtype             = 'CartridgeItem';
-   static public $items_id             = 'cartridgeitems_id';
+   static public $itemtype_1 = 'CartridgeItem';
+   static public $items_id_1 = 'cartridgeitems_id';
+
+   static public $itemtype_2 = 'Printer';
+   static public $items_id_2 = 'printers_id';
 
    public function getCloneRelations() :array {
       return [
```

**Comment by M-Falken** [comment]
Hi
it fix the issue for me, but no side effects seen for me at this moment.
So does this part:
```
+   static public $itemtype_1 = 'CartridgeItem';
+   static public $items_id_1 = 'cartridgeitems_id';
+
+   static public $itemtype_2 = 'Printer';
+   static public $items_id_2 = 'printers_id';
```
will be added in the actual 9.5 bugfix or in the next one with 9.5.4 ?
Thanks

**Comment by eduardomozart** [comment]
Hello @cedric-anne 
I applied your fix into our GLPI 9.5.4 but the issue of #8108 (linked by @cconard96 into this issue) still persists. It's a problem that is happening since GLPI 9.5.2, at least. I believe it's happening because, as stated by @cconard96, GLPI uses the same code for template and copying now.

**Comment by M-Falken** [comment]
Hello,
still the same issue with 9.5.5
`[2021-04-22 13:24:56] glpiphplog.CRITICAL:   *** Uncaught Exception RuntimeException: Cannot guess field for itemtype Printer on Cartridge in /home/public_html/inc/commondbchild.class.php at line 882
  Backtrace :
  inc/features/clonable.class.php:69                 CommonDBChild::getItemField()
  inc/commondbtm.class.php:1264                      Printer->post_clone()
  inc/commondbtm.class.php:1092                      CommonDBTM->clone()
  front/printer.form.php:50                          CommonDBTM->add()`

**Comment by chnateag** [comment]
Hi guys, applied this patch, it worked fine but i've got an new issue when i want to add a cartridge : 
`Uncaught Exception Error: Access to undeclared static property: Cartridge::$itemtype in /home/www/glpi/inc/cartridge.class.php at line 99`

**Comment by mathieumd** [comment]
The commit c50d72300d9c1f6fe2f87f9834a570cce52c4e5d fixes the problem for me too, on my GLPI 9.5.5.

Why [hasn't it been included in 9.5.6](https://github.com/glpi-project/glpi/blob/9.5.6/inc/printer.class.php#L63)? :-(

**Comment by trasher** [comment]
> The commit [c50d723](https://github.com/glpi-project/glpi/commit/c50d72300d9c1f6fe2f87f9834a570cce52c4e5d) fixes the problem for me too, on my GLPI 9.5.5.
> 
> Why [hasn't it been included in 9.5.6](https://github.com/glpi-project/glpi/blob/9.5.6/inc/printer.class.php#L63)? :-(

https://github.com/glpi-project/glpi/pull/8984#issuecomment-824824498

**Comment by mathieumd** [comment]
I'm not sure what is really the problem highlighed by this log. Indeed I do get too the PHP error log `Unable to clone elements of class CartridgeItem as it does not extends "CommonDBConnexity"` when I clone a printer, but it's still cloned without apparent problem. Cartridges are no more linked, but it could be the normal behaviour (I never tried to clone printers before).

**Comment by cedric-anne** [comment]
Hi,

#9558 is an alternative solution. Could you test it ?

Regards

**Comment by mathieumd** [comment]
I applied [changed listed there](https://github.com/glpi-project/glpi/pull/9558/files) to files of 9.5.5, and I confirm I can add a printer from template, and then clone that freshly added printer without any error loggued in `php-errors.log`! Nice! :)

**Comment by cedric-anne** [comment]
Fixed by #9558 .
